File: Controlador/ControladorPrincipal.cs
using System;
using System.Collections.Generic;
using punto_de_venta_C_.Modelo;
using punto_de_venta_C_.ConexionBD;
using punto_de_venta_C_.Vista;

namespace punto_de_venta_C_.Controlador
{
    public class ControladorPrincipal
    {
        private readonly UsuarioBD usuarioBD = new UsuarioBD();
        private readonly ProductoBD productoBD = new ProductoBD();
        private readonly VentaBD ventaBD = new VentaBD();
        private readonly MovimientoBD movimientoBD = new MovimientoBD();
        private readonly ProveedorBD proveedorBD = new ProveedorBD();
        private readonly OrdenCompraBD ordenBD = new OrdenCompraBD();
        private readonly CuentaPorPagarBD cuentaBD = new CuentaPorPagarBD();
        private readonly DevolucionBD devolucionBD = new DevolucionBD();
        private readonly ConfiguracionBD configBD = new ConfiguracionBD();

        private double montoCaja = 0;
        private bool cajaAbierta = false;
        private Usuario usuarioActivo;

        private ConfiguracionTienda config = new ConfiguracionTienda("CORPORATIVO POS", "Sucursal Principal - Centro", "XAXX010101000");

        private LoginVista loginVista;
        private VentanaPrincipal ventanaPrincipal;

        private VentaControlador ventaCtrl;
        private InventarioControlador inventarioCtrl;
        private CajaControlador cajaCtrl;
        private ProveedorControlador proveedorCtrl;
        private UsuarioControlador usuarioCtrl;

        public void iniciar()
        {
            ConfiguracionTienda guardada = configBD.obtener();
            if (guardada != null)
            {
                this.config = guardada;
            }

            cargarProductosIniciales();

            loginVista = new LoginVista(this);
            loginVista.mostrar();
        }

        public void onLoginExitoso(Usuario u)
        {
            this.usuarioActivo = u;
            loginVista.ocultar();
            ventaCtrl = new VentaControlador(this);
            inventarioCtrl = new InventarioControlador(this);
            cajaCtrl = new CajaControlador(this);
            proveedorCtrl = new ProveedorControlador(this);
            usuarioCtrl = new UsuarioControlador(this);
            ventanaPrincipal = new VentanaPrincipal(this, ventaCtrl, inventarioCtrl, cajaCtrl, proveedorCtrl, usuarioCtrl);
            ventanaPrincipal.mostrar();
        }

        public void onCerrarSesion()
        {
            this.usuarioActivo = null;
            this.cajaAbierta = false;
            this.montoCaja = 0;
            if (ventanaPrincipal != null)
                ventanaPrincipal.ocultar();
            loginVista.limpiar();
            loginVista.mostrar();
        }

        public Usuario autenticar(string username, string password)
        {
            return usuarioBD.autenticar(username, password);
        }

        public void registrarVenta(Venta venta)
        {
            ventaBD.insertar(venta);

            if (venta.getMetodoPago().Equals("Efectivo", StringComparison.OrdinalIgnoreCase))
            {
                this.montoCaja += venta.getTotal();
            }

            movimientoBD.insertar(
                new Movimiento(Movimiento.Tipo.VENTA, "venta #" + venta.getId() + " (" + venta.getMetodoPago() + ")",
                        venta.getTotal(), DateTime.Now, usuarioActivo.getNombre()));

            if (ventanaPrincipal != null)
            {
                ventanaPrincipal.refrescarCaja();
                ventanaPrincipal.refrescarInventario();
            }
        }

        public void registrarDevolucion(Devolucion d)
        {
            devolucionBD.insertar(d);
            Producto p = productoBD.buscarPorId(d.getProducto().getId());
            if (p != null)
                productoBD.actualizarStock(p.getId(), p.getStock() + d.getCantidad());
            movimientoBD.insertar(new Movimiento(Movimiento.Tipo.RETIRO, "Devolucion Venta #" + d.getVentaId(),
                    d.getMontoDevuelto(), DateTime.Now, usuarioActivo.getNombre()));
            if (ventanaPrincipal != null)
                ventanaPrincipal.refrescarInventario();
        }

        public void registrarVentaSimple(double monto)
        {
            this.montoCaja += monto;
            if (ventanaPrincipal != null)
            {
                ventanaPrincipal.refrescarCaja();
            }
        }

        public void registrarRetiro(string concepto, double monto)
        {
            this.montoCaja -= monto;
            movimientoBD.insertar(new Movimiento(Movimiento.Tipo.RETIRO, concepto, monto, DateTime.Now,
                    usuarioActivo.getNombre()));
            if (ventanaPrincipal != null)
            {
                ventanaPrincipal.refrescarCaja();
            }
        }

        public void abrirCaja(double fondo)
        {
            this.montoCaja = fondo;
            this.cajaAbierta = true;

            movimientoBD.insertar(new Movimiento(Movimiento.Tipo.VENTA, "apertura de caja", fondo, DateTime.Now,
                    usuarioActivo.getNombre()));

            if (ventanaPrincipal != null)
            {
                ventanaPrincipal.refrescarCaja();
            }
        }

        public void guardarConfiguracion(ConfiguracionTienda nuevaConfig)
        {
            this.config = nuevaConfig;

            configBD.actualizar(nuevaConfig);

            if (ventanaPrincipal != null)
            {
                ventanaPrincipal.actualizarTitulo();
            }
        }

        public void reiniciarSistemaCompleto()
        {
            try
            {
                string[] tablas = { "devoluciones", "cuentas_por_pagar", "ordenes_compra", "movimientos", "ventas" };

                using (MySqlConnector.MySqlConnection c = Conexion.getConexion())
                {
                    using (MySqlConnector.MySqlCommand st = c.CreateCommand())
                    {
                        st.CommandText = "SET FOREIGN_KEY_CHECKS = 0";
                        st.ExecuteNonQuery();

                        foreach (string tabla in tablas)
                        {
                            st.CommandText = "TRUNCATE TABLE " + tabla;
                            st.ExecuteNonQuery();
                        }

                        st.CommandText = "SET FOREIGN_KEY_CHECKS = 1";
                        st.ExecuteNonQuery();
                    }
                }

                this.montoCaja = 0;
                this.cajaAbierta = false;

                if (ventanaPrincipal != null)
                {
                    ventanaPrincipal.refrescarCaja();
                    this.onCerrarSesion();
                }

                if (Microsoft.Maui.Controls.Application.Current?.MainPage != null)
                    Microsoft.Maui.Controls.Application.Current.MainPage.DisplayAlert("Sistema", "sistema reiniciado con exito", "OK");

            }
            catch (Exception e)
            {
                if (Microsoft.Maui.Controls.Application.Current?.MainPage != null)
                    Microsoft.Maui.Controls.Application.Current.MainPage.DisplayAlert("Error", "error al reiniciar: " + e.Message, "OK");
            }
        }

        public List<Usuario> getUsuarios() { return usuarioBD.obtenerTodos(); }
        public List<Producto> getProductos() { return productoBD.obtenerTodos(); }
        public List<Venta> getVentas() { return ventaBD.obtenerTodas(); }
        public List<Movimiento> getMovimientos() { return movimientoBD.obtenerDelDia(); }
        public List<Proveedor> getProveedores() { return proveedorBD.obtenerTodos(); }
        public List<OrdenCompra> getOrdenes() { return ordenBD.obtenerTodas(); }
        public List<CuentaPorPagar> getCuentas() { return cuentaBD.obtenerActivas(); }
        public List<Devolucion> getDevoluciones() { return devolucionBD.obtenerTodas(); }

        public UsuarioBD getUsuarioBD() { return usuarioBD; }
        public ProductoBD getProductoBD() { return productoBD; }
        public VentaBD getVentaBD() { return ventaBD; }
        public MovimientoBD getMovimientoBD() { return movimientoBD; }
        public ProveedorBD getProveedorBD() { return proveedorBD; }
        public OrdenCompraBD getOrdenBD() { return ordenBD; }
        public CuentaPorPagarBD getCuentaBD() { return cuentaBD; }
        public DevolucionBD getDevolucionBD() { return devolucionBD; }

        public ConfiguracionTienda getConfig() { return config; }
        public void setConfig(ConfiguracionTienda c) { this.config = c; }
        public double getMontoCaja() { return montoCaja; }
        public bool isCajaAbierta() { return cajaAbierta; }
        public Usuario getUsuarioActivo() { return usuarioActivo; }
        public VentanaPrincipal getVentanaPrincipal() { return ventanaPrincipal; }
        
        private void cargarProductosIniciales()
        {
            try
            {
                List<Producto> productosBase = DatosIniciales.getProductos();
                int insertados = 0;
                int actualizados = 0;
                foreach (Producto p in productosBase)
                {
                    Producto existente = productoBD.buscarPorId(p.getId());
                    if (existente == null)
                    {
                        productoBD.insertar(p);
                        insertados++;
                    }
                    else
                    {
                        if (existente.getNombre() != p.getNombre() || 
                            existente.getPrecio() != p.getPrecio() || 
                            existente.getCategoria() != p.getCategoria())
                        {
                            p.setStock(existente.getStock());
                            productoBD.actualizar(p);
                            actualizados++;
                        }
                    }
                }
                Console.WriteLine("Seed finalizado. Insertados: " + insertados + ", Actualizados: " + actualizados);
            }
            catch (Exception e)
            {
                Console.Error.WriteLine("Error en seed de productos: " + e.Message);
            }
        }
    }
}

File: Controlador/ControladorPrincipal.py
from ConexionBD.UsuarioBD import UsuarioBD
from ConexionBD.ProductoBD import ProductoBD
from ConexionBD.VentaBD import VentaBD
from ConexionBD.MovimientoBD import MovimientoBD
from ConexionBD.ProveedorBD import ProveedorBD
from ConexionBD.OrdenCompraBD import OrdenCompraBD
from ConexionBD.CuentaPorPagarBD import CuentaPorPagarBD
from ConexionBD.DevolucionBD import DevolucionBD
from ConexionBD.ConfiguracionBD import ConfiguracionBD
from Modelo.ConfiguracionTienda import ConfiguracionTienda
from Modelo.Movimiento import Movimiento, Tipo as TipoMovimiento
from datetime import datetime
from PyQt6.QtWidgets import QMessageBox

class ControladorPrincipal:
    def __init__(self):
        self.usuarioBD = UsuarioBD()
        self.productoBD = ProductoBD()
        self.ventaBD = VentaBD()
        self.movimientoBD = MovimientoBD()
        self.proveedorBD = ProveedorBD()
        self.ordenBD = OrdenCompraBD()
        self.cuentaBD = CuentaPorPagarBD()
        self.devolucionBD = DevolucionBD()
        self.configBD = ConfiguracionBD()

        self.montoCaja = 0.0
        self.cajaAbierta = False
        self.usuarioActivo = None

        self.config = ConfiguracionTienda()
        self.config.setNombre("CORPORATIVO POS")
        self.config.setSucursal("Sucursal Principal - Centro")
        self.config.setRfc("XAXX010101000")

        self.loginVista = None
        self.ventanaPrincipal = None

        self.ventaCtrl = None
        self.inventarioCtrl = None
        self.cajaCtrl = None
        self.proveedorCtrl = None
        self.usuarioCtrl = None

    def iniciar(self):
        guardada = self.configBD.obtener()
        if guardada is not None:
            self.config = guardada

        self.cargarProductosIniciales()

        from Vista.LoginVista import LoginVista
        self.loginVista = LoginVista(self)
        self.loginVista.mostrar()

    def onLoginExitoso(self, u):
        self.usuarioActivo = u
        if self.loginVista:
            self.loginVista.ocultar()
        
        from Controlador.VentaControlador import VentaControlador
        from Controlador.InventarioControlador import InventarioControlador
        from Controlador.CajaControlador import CajaControlador
        from Controlador.ProveedorControlador import ProveedorControlador
        from Controlador.UsuarioControlador import UsuarioControlador
        from Vista.VentanaPrincipal import VentanaPrincipal
        
        self.ventaCtrl = VentaControlador(self)
        self.inventarioCtrl = InventarioControlador(self)
        self.cajaCtrl = CajaControlador(self)
        self.proveedorCtrl = ProveedorControlador(self)
        self.usuarioCtrl = UsuarioControlador(self)
        
        self.ventanaPrincipal = VentanaPrincipal(self, self.ventaCtrl, self.inventarioCtrl, self.cajaCtrl, self.proveedorCtrl, self.usuarioCtrl)
        self.ventanaPrincipal.mostrar()

    def onCerrarSesion(self):
        self.usuarioActivo = None
        self.cajaAbierta = False
        self.montoCaja = 0.0
        if self.ventanaPrincipal is not None:
            self.ventanaPrincipal.ocultar()
        if self.loginVista:
            self.loginVista.limpiar()
            self.loginVista.mostrar()

    def autenticar(self, username, password):
        return self.usuarioBD.autenticar(username, password)

    def registrarVenta(self, venta):
        self.ventaBD.insertar(venta)
        if str(venta.getMetodoPago()).lower() == "efectivo":
            self.montoCaja += venta.getTotal()

        m = Movimiento()
        m.setTipo(TipoMovimiento.VENTA)
        m.setDescripcion(f"venta #{venta.getId()} ({venta.getMetodoPago()})")
        m.setMonto(venta.getTotal())
        m.setFecha(datetime.now())
        m.setUsuario(self.usuarioActivo.getNombre() if self.usuarioActivo else "SISTEMA")
        self.movimientoBD.insertar(m)

        if self.ventanaPrincipal is not None:
            self.ventanaPrincipal.refrescarCaja()
            self.ventanaPrincipal.refrescarInventario()

    def registrarDevolucion(self, d):
        self.devolucionBD.insertar(d)
        p = self.productoBD.buscarPorId(d.getProducto().getId())
        if p is not None:
            self.productoBD.actualizarStock(p.getId(), p.getStock() + d.getCantidad())
            
        m = Movimiento()
        m.setTipo(TipoMovimiento.RETIRO)
        m.setDescripcion(f"Devolucion Venta #{d.getVentaId()}")
        m.setMonto(d.getMontoDevuelto() if hasattr(d, 'getMontoDevuelto') else 0.0) # Assuming getMontoDevuelto exists or 0
        m.setFecha(datetime.now())
        m.setUsuario(self.usuarioActivo.getNombre() if self.usuarioActivo else "SISTEMA")
        self.movimientoBD.insertar(m)
        
        if self.ventanaPrincipal is not None:
            self.ventanaPrincipal.refrescarInventario()

    def registrarVentaSimple(self, monto):
        self.montoCaja += monto
        if self.ventanaPrincipal is not None:
            self.ventanaPrincipal.refrescarCaja()

    def registrarRetiro(self, concepto, monto):
        self.montoCaja -= monto
        m = Movimiento()
        m.setTipo(TipoMovimiento.RETIRO)
        m.setDescripcion(concepto)
        m.setMonto(monto)
        m.setFecha(datetime.now())
        m.setUsuario(self.usuarioActivo.getNombre() if self.usuarioActivo else "SISTEMA")
        self.movimientoBD.insertar(m)
        
        if self.ventanaPrincipal is not None:
            self.ventanaPrincipal.refrescarCaja()

    def abrirCaja(self, fondo):
        self.montoCaja = fondo
        self.cajaAbierta = True

        m = Movimiento()
        m.setTipo(TipoMovimiento.VENTA)
        m.setDescripcion("apertura de caja")
        m.setMonto(fondo)
        m.setFecha(datetime.now())
        m.setUsuario(self.usuarioActivo.getNombre() if self.usuarioActivo else "SISTEMA")
        self.movimientoBD.insertar(m)

        if self.ventanaPrincipal is not None:
            self.ventanaPrincipal.refrescarCaja()

    def guardarConfiguracion(self, nuevaConfig):
        self.config = nuevaConfig
        self.configBD.actualizar(nuevaConfig)
        if self.ventanaPrincipal is not None:
            self.ventanaPrincipal.actualizarTitulo()

    def reiniciarSistemaCompleto(self):
        try:
            tablas = ["devoluciones", "cuentas_por_pagar", "ordenes_compra", "movimientos", "ventas"]
            from ConexionBD.Conexion import Conexion
            conn = Conexion.getConexion()
            if conn:
                cursor = conn.cursor()
                cursor.execute("SET FOREIGN_KEY_CHECKS = 0")
                for tabla in tablas:
                    cursor.execute(f"TRUNCATE TABLE {tabla}")
                cursor.execute("SET FOREIGN_KEY_CHECKS = 1")
                conn.commit()
                cursor.close()

            self.montoCaja = 0.0
            self.cajaAbierta = False

            if self.ventanaPrincipal is not None:
                self.ventanaPrincipal.refrescarCaja()
                self.onCerrarSesion()

            QMessageBox.information(None, "Éxito", "sistema reiniciado con exito")
        except Exception as e:
            QMessageBox.critical(None, "Error", f"error al reiniciar: {e}")

    def getUsuarios(self):
        return self.usuarioBD.obtenerTodos()

    def getProductos(self):
        return self.productoBD.obtenerTodos()

    def getVentas(self):
        return self.ventaBD.obtenerTodas()

    def getMovimientos(self):
        return self.movimientoBD.obtenerDelDia()

    def getProveedores(self):
        return self.proveedorBD.obtenerTodos()

    def getOrdenes(self):
        return self.ordenBD.obtenerTodas()

    def getCuentas(self):
        return self.cuentaBD.obtenerActivas()

    def getDevoluciones(self):
        return self.devolucionBD.obtenerTodas()

    def getUsuarioBD(self):
        return self.usuarioBD

    def getProductoBD(self):
        return self.productoBD

    def getVentaBD(self):
        return self.ventaBD

    def getMovimientoBD(self):
        return self.movimientoBD

    def getProveedorBD(self):
        return self.proveedorBD

    def getOrdenBD(self):
        return self.ordenBD

    def getCuentaBD(self):
        return self.cuentaBD

    def getDevolucionBD(self):
        return self.devolucionBD

    def getConfig(self):
        return self.config

    def setConfig(self, c):
        self.config = c

    def getMontoCaja(self):
        return self.montoCaja

    def isCajaAbierta(self):
        return self.cajaAbierta

    def getUsuarioActivo(self):
        return self.usuarioActivo

    def getVentanaPrincipal(self):
        return self.ventanaPrincipal

    def cargarProductosIniciales(self):
        try:
            from Modelo.DatosIniciales import DatosIniciales
            productosBase = DatosIniciales.getProductos()
            if not productosBase: return
            insertados = 0
            actualizados = 0
            for p in productosBase:
                existente = self.productoBD.buscarPorId(p.getId())
                if existente is None:
                    self.productoBD.insertar(p)
                    insertados += 1
                else:
                    if existente.getNombre() != p.getNombre() or existente.getPrecio() != p.getPrecio() or existente.getCategoria() != p.getCategoria():
                        p.setStock(existente.getStock())
                        self.productoBD.actualizar(p)
                        actualizados += 1
            print(f"Seed finalizado. Insertados: {insertados}, Actualizados: {actualizados}")
        except Exception as e:
            print(f"Error en seed de productos: {e}")
